### PR TITLE
Pedantic alignment checks: make safe stuff truly safe

### LINKED ISSUE #35: Not safe: Performs unaligned memory accesses
The following code:
```rust
extern crate safe_transmute;

fn main() {
    let bytes: &[u8] = &[0, 1, 2, 3, 4, 5];
    println!("{}", safe_transmute::guarded_transmute_pod::<u32>(bytes).unwrap());
    println!("{}", safe_transmute::guarded_transmute_pod::<u32>(&bytes[1..]).unwrap());
}
```
Creates an unaligned reference, and, worse, accesses memory through it. Both of these (even just the creation!) are undefined behavior. In other words, this crate exports functions safe that are not safe to use.

Generally, the entire safe API of this crate seems to be built ignoring alignment constraints. I am afraid that's *not safe*.

FWIW, [miri](https://github.com/solson/miri/) finds this error and shows
```
error[E0080]: constant evaluation error: type validation failed: encountered unaligned reference
    --> /home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/slice/mod.rs:4789:5
     |
4789 |     Repr { raw: FatPtr { data, len } }.rust
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type validation failed: encountered unaligned reference
     |
     = note: inside call to `std::slice::from_raw_parts::<u32>` at /home/r/.cargo/registry/src/github.com-1ecc6299db9ec823/safe-transmute-0.10.0/src/lib.rs:151:8
     = note: inside call to `safe_transmute::guarded_transmute::<u32>` at /home/r/.cargo/registry/src/github.com-1ecc6299db9ec823/safe-transmute-0.10.0/src/pod.rs:100:14
note: inside call to `safe_transmute::guarded_transmute_pod::<u32>` at src/main.rs:5:20
    --> src/main.rs:5:20
     |
5    |     println!("{}", safe_transmute::guarded_transmute_pod::<u32>(bytes).unwrap());
     |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

### PR DESCRIPTION
**Pull request by Enet4**
**tl;dr make `pod_*` functions raise errors on bad alignment, we'll make its use more fluid in the future**

~~**Please merge #36 first, and consider releasing a patch once it is (merged).**~~

With this PR, we aim to make non-`unsafe` things truly memory safe by preventing a conversion from happening if the memory cannot be read with the memory alignment expected by the destination type. This applies to just about all functions in the `pod` module. The tests were modified to ensure proper memory alignment (except for the doctests, which are only compiled for the time being because that would make them a bit more complicated).

Generally, this means that all of these functions can arbitrarily fail in a way that a typical user might not predict at first, and might be hard to. In fact, one cannot usually ensure the intended alignment, especially when working with vectors. The alignment of this one is chosen by the heap allocator, and the `Vec` API does not expose such allocation details (which is a good thing!).

However, future plans for the crate would then be to provide ways to overcome this issue. One of these is to make it easy to produce a copy if a conversion is not possible. Another one would be new helper functions to construct well-aligned data vectors. The function `aligned_vec`, for example, only exists for testing purposes, but could become part of the public API once it's well documented and well tested. More decisions after this is merged will follow.

Summary of changes:

- Added `util::check_alignment()` and respective unit tests;
- A call `util::check_alignment()` was added to all functions in `pod`;
- The `pod` module and functions were documented with all this;
- Added `aligned_vec(&[u8])` in the integration tests: it creates a copy of the given bytes into a `Vec<u8>` that is sure to be aligned for safe access if it's transmuted to a slice or vector of `T`, but this guarantee is lost if the vector is modified in certain ways (adding/removing elements, clearing, reserving, etc.);
- Tests were updated to use `aligned_vec()` or `guarded_transmute_to_bytes_pod_many()` to create well aligned memory. By turning a slice of words into a slice of bytes, we get properly aligned memory for free and no longer need to account for byte order.

Resolves #35, although some tests using the unsafe function variants still assume that the inputs' alignment is fine. I kept it this way because it works as intended for the time being, but it might be worth changing those too eventually.

### CODE REVIEW

File: src/util.rs [hunk 1]
```diff
@@ -63,3 +63,31 @@ pub fn check_alignment<T, U>(data: &[T]) -> Result<(), Error> {
         Ok(())
     }
 }
+
+#[cfg(test)]
```
[the remark this answers is not in the document]
Comment (Enet4) [reply src/util.rs]: I kept it here because it's a unit test rather than an integration test. But OK, will do.

File: tests/alignment_check.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+#[cfg(feature = "std")]
+extern crate core;
+extern crate safe_transmute;
+
+use safe_transmute::guarded_transmute_pod_many;
+use safe_transmute::guarded_transmute_to_bytes_pod_many;
+use safe_transmute::Error;
+
+#[test]
+fn unaligned_slicing_integers() {
```
[the remark this answers is not in the document]
Comment (Enet4) [reply tests/alignment_check.rs]: Hmm. I would not call having multiple test binaries an issue, but I'll try to do as you ask.

File: src/pod.rs [hunk 3]
```diff
@@ -1,7 +1,15 @@
+//! Plain data object safe transmute
```
[the remark this answers is not in the document]
Comment (Enet4) [reply src/pod.rs]: Aye, I know. Not yet. #9 :P

File: tests/test_util/aligned_vec.rs [hunk 4]
```diff
@@ -1,3 +1,4 @@
+#[cfg(feature = "std")]
```
Comment (nabijaczleweli) [inline tests/test_util/aligned_vec.rs]: Not ` #![cfg(feature = "std")] `?

**Comment by Enet4** [comment]
All changes are made. Note that Clippy doth protest too much methinks, and should now be installed via Rustup.

File: tests/lib.rs [hunk 5]
```diff
@@ -4,6 +4,11 @@ extern crate core;
 
 extern crate safe_transmute;
 
+#[cfg(feature = "std")]
```
Comment (nabijaczleweli) [inline tests/lib.rs]: My only question for this commit (df5346627b3ee712e0fbbfff8991b32d63ad228b) in its entirety is: why?

Comment (Enet4) [reply tests/lib.rs]: Because `include!`s are a mess. If I don't centralise these `use`s to a single place, there will be warnings or errors. Perhaps we should get rid of them entirely and make it an actual module?

Comment (nabijaczleweli) [reply tests/lib.rs]: Can't make it an "actual module" because we're including it in doctests. argh, schite.

Allllllllllllllllllllllllllllllllllllllllllllllllllrighty then, as much as I'm pained to do this, the only reasonable-ish solution *is* to move the `use`s into the funxion bodies, innit :v

Comment (Enet4) [reply tests/lib.rs]: That was a bit unclear. Do you want me to move the `use`s then?

Comment (nabijaczleweli) [reply tests/lib.rs]: Probably, if there's no better way, aye

Comment (Enet4) [reply tests/lib.rs]: It is done in 1d79fa6.

**Comment by nabijaczleweli** [comment]
Magnificent, tremendous, only two and a half weeks. I must say, we're nearing a record!
